pilot/pkg/networking/plugin/authz/rbac_v2.go

```diff
@@ -0,0 +1,172 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package authz converts Istio RBAC (role-based-access-control) policies (ServiceRole and AuthorizationPolicy)
+// to corresponding filter config that is used by the envoy RBAC filter to enforce access control to
+// the service co-located with envoy.
+// Currently the config is only generated for sidecar node on inbound HTTP/TCP listener. The generation
+// is controlled by RbacConfig (a singleton custom resource with cluster scope). User could disable
+// this plugin by either deleting the ClusterRbacConfig or set the ClusterRbacConfig.mode to OFF.
+// Note: ClusterRbacConfig is not created with default istio installation which means this plugin doesn't
+// generate any RBAC config by default.
+package authz
+
+import (
+	http_config "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/rbac/v2"
+	policyproto "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2alpha"
+
+	rbacproto "istio.io/api/rbac/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// convertRbacRulesToFilterConfigV2 is the successor of convertRbacRulesToFilterConfig, which supports
+// converting AuthorizationPolicy.
+func convertRbacRulesToFilterConfigV2(service *serviceMetadata, option rbacOption) *http_config.RBAC {
+	rbac := &policyproto.RBAC{
+		Action:   policyproto.RBAC_ALLOW,
+		Policies: map[string]*policyproto.Policy{},
+	}
+
+	if option.globalPermissiveMode {
+		// Permissive Mode will be ignored.
+		// TODO(pitlv2109): Handle permissive mode in the future.
+		return &http_config.RBAC{Rules: rbac}
+	}
+
+	namespace := service.attributes[attrDestNamespace]
+
+	// Get all AuthorizationPolicy Istio config from this namespace.
+	allAuthzPolicies := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace].AuthzPolicies
+	// Get all key-value pairs of AuthorizationPolicy name to list of Subjects/ServiceRoleBindings in this namespace.
+	authzPolicyToSubjects := option.authzPolicies.AuthzPolicyToAllowSubjects(namespace)
+	for _, authzPolicy := range allAuthzPolicies {
+		// If a WorkloadSelector is used in the AuthorizationPolicy config and does not match this service, skip this
+		// AuthorizationPolicy.
+		if authzPolicy.Policy.WorkloadSelector != nil && !service.isServiceMatchedWorkload(authzPolicy.Policy.WorkloadSelector.Labels) {
+			continue
+		}
+		for i, subject := range authzPolicyToSubjects[authzPolicy.Name] {
+			// Convert role.
+			role := option.authzPolicies.GetServiceRoleFromName(namespace, subject.RoleRef.Name)
+			permission := &policyproto.Permission{}
+			rule := role.Rules[0]
+
+			// Check to make sure that the current service (caller) is the one this rule is applying to.
+			if !service.match(rule) {
```

Comment:
In authz-v2, A `ServiceRole` shouldn't use service field anymore (replaced by label selector in authorization policy), so do a double check before here and log an error message if it happens.

Reply:
The design doc mentions that the label selector is optional. So for each service, we still need to check if the current ServiceRole applies to that service?

Reply:
Yes, but not only because label selector is optional, ServiceRole include other fields that need to match in its constraints.